lib/python/picongpu/extra/input/preparingInsightData.py

```diff
@@ -0,0 +1,472 @@
+        b = np.fft.fftfreq(self.y.size, d=self.dy)
+        B, A, W = np.meshgrid(b, a, self.w)
+        cond = (2 * np.pi * c / W) ** 2 * (A**2 + B**2)
+        idx_in = np.where(cond < 1)
+        idx_out = np.where(cond > 1)
+        F_Ew = np.fft.fft2(self.Ew, axes=(0, 1))
+        F_Ew[idx_in] *= np.exp(1j * W[idx_in] / c * np.sqrt(1 - cond[idx_in]) * z)
+        F_Ew[idx_out] *= np.exp(-W[idx_out] / c * np.sqrt(cond[idx_out] - 1) * z)
+        lin_phase = np.exp(-1j * W * z / c)  # cancel out the time shift
+
+        print("far field propagated to z = %.2f mm" % (z))
+        return np.fft.ifft2(F_Ew, axes=(0, 1)) * lin_phase
+
+    def toTimeDomain(self, Ew, lamb_supp=10):
+        """
+        Transform far field from frequency to time domain
+
+        Arguments:
+        Ew: far field data in frequency domain (at the focus or propagated)
+        lamb_supp: number of sampling points on one wavelength;
+                   from this, the number of zeros to be padded will be derived. Default is 10.
+        """
+        assert self.isPhaseCorrected, "Oops, you forgot to correct the phase!"
+
+        # calculate the number of zeros to be padded at the right side of the spectrum
+        # longitudinal axis length N_tot = 2*pi/(dw*dt)
+        # necessary time spacing dt = lambda/(c*lamb_supp)
+        zeros = int(self.w[self.wc_idx] * lamb_supp / self.dw / 2 - self.w[-1] / self.dw)
+
+        # the data has to be sorted according to fft requirements:  w = dw* (0, 1, ... n/2, -n/2, ... -1)
+        # so we have to extent (and interpolate) first the positive side of the spectrum
+        w_fft = np.arange(0, self.w[-1] + zeros * self.dw, self.dw)
+        Y_fft, X_fft, W_fft = np.meshgrid(self.y, self.x, w_fft, indexing="ij")
+        fft_interp = RegularGridInterpolator(
+            (self.y, self.x, self.w), Ew, bounds_error=False, fill_value=0, method="linear"
+        )
+        Ew_fft = fft_interp((Y_fft, X_fft, W_fft))
+        # set the negative frequency part to 0 (instead of the complex conjugate) and neglegt the imaginary part afterwards
+        w_fft = np.concatenate((w_fft, -np.flip(w_fft[1:])))
+        Ew_fft = np.concatenate((Ew_fft, np.zeros_like(Ew_fft[:, :, :-1])), axis=-1)
+
+        # transform to time domain
+        self.Et = np.fft.fftshift(np.fft.ifft(Ew_fft), axes=-1)
+        self.Et = self.Et / np.abs(self.Et).max()  # rescale amplitude to one
+        self.t = np.fft.fftshift(np.fft.fftfreq(w_fft.size, self.dw / (2 * np.pi)))  # fs
+        self.dt = np.diff(self.t)[0]
+
+        print("field data size: %.1f GB" % (np.prod(self.Et.shape) * 8 * 10**-9))  # assume datatype double
+
+    def saveToOpenPMD(self, outputpath, outputname, energy, pol="x", crop_x=0, crop_y=0, crop_t=0):
+        """
+        Save the field data in time domain to an OpenPMD checkpoint. This output will be used for the InsightPulse profile.
+        The propagation axis ("z") will be transformed via z = c*t to a spatial one.
```

Comment:
That is true. This transformation step is nonsense from a physical poin of view, it is just done to fulfil the openPMD requirements to store a field chunk (in just one iteration) and undone stright away again by the FromOpenPMDPulse profile. Certainly it would be more correct to store each time slice in their respective iteration, but since there are around 1000 of them per chunk, it will not only complicate the source code but probably also increase the computation time (I guess, didn't try it out yet). 
Nevertheless, I added it to the refactoring list.